src/lib.rs

```diff
@@ -365,6 +427,7 @@ pub mod map;
 pub mod ser;
```

Comment:
From `cargo doc --no-default-features --features alloc`:

> ![Screenshot from 2020-01-20 23-26-49](https://user-images.githubusercontent.com/1940490/72783907-60236e00-3bdc-11ea-8062-2288dd6bd35c.png)

<br>

This seems bad because the following would be accepted in alloc mode and break if a different part of the dependency graph enables std mode.

```rust
struct S;
impl serde_json::ser::Formatter for S {
    fn write_null<W: ?Sized>(&mut self, _: &mut W) -> Result<(), &'static str> {
        Err("")
    }
}
```

Either Formatter needs to be removed from the public API in alloc mode or the error type needs to be changed to an opaque type with no more than a subset of io::Error's API.

Reply:
Oops, I didn't notice that, thanks!

Since `std::io::Result` is part of the default, public API I don't think we can be backwards compatible (methods are one thing but sth may depend on it being explicitly this type from `std` IIUC).

Because of this, I think the only thing we can do now, until `std::io` somehow lands in `core`, is to hide this from public API in no-`std` case.

Reply:
Reimplemented a small subset of `std::io::Error` and friends so that accidentally opting into `std` should not break stuff